File: core/src/main/java/org/tgoe/core/member/beans/Member.java
package org.tgoe.core.member.beans;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tgoe.core.member.enums.MemberDosbGender;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

@JsonIgnoreProperties(ignoreUnknown = true)
public class Member {
	public final static String easyvereinQueryString = "{id,membershipNumber,joinDate,resignationDate,emailOrUserName,_isChairman,"
			+ "contactDetails{salutation,firstName,familyName,dateOfBirth,street,city,zip,privateEmail},"
			+ "memberGroups{memberGroup" + MemberGroup.easyvereinQueryString + "},"
			+ "integrationDosbSport,integrationDosbGender}";

	private final SimpleDateFormat dateOnlyFormat = new SimpleDateFormat("yyyy-MM-dd");
	private static final Logger logger = LoggerFactory.getLogger(Member.class);

	private long id;
	private String membershipNumber;
	private Date joinDate;
	private Date resignationDate;
	private String loginName;
	private boolean isAdmin;
	private String salutation;
	private String firstName;
	private String familyName;
	private Date dateOfBirth;
	private String street;
	private String city;
	private String zip;
	private String privateEmail;
	private MemberDosbGender dosbGender;
	private List<GroupMembership> groupMemberships;

	public long getId() {
		return id;
	}

	@JsonProperty("id")
	private void setId(long id) {
		this.id = id;
	}

	public String getMembershipNumber() {
		return membershipNumber;
	}

	@JsonProperty("membershipNumber")
	private void setMembershipNumber(String membershipNumber) {
		this.membershipNumber = membershipNumber;
	}

	public Date getJoinDate() {
		return joinDate;
	}

	@JsonProperty("joinDate")
	private void setJoinDate(Date joinDate) {
		this.joinDate = joinDate;
	}

	public Date getResignationDate() {
		return resignationDate;
	}

	@JsonProperty("resignationDate")
	private void setResignationDate(Date resignationDate) {
		this.resignationDate = resignationDate;
	}

	public String getLoginName() {
		return loginName;
	}

	@JsonProperty("emailOrUserName")
	private void setLoginName(String loginName) {
		this.loginName = loginName;
	}

	public boolean isAdmin() {
		return isAdmin;
	}

	@JsonProperty("_isChairman")
	private void setAdmin(boolean isAdmin) {
		this.isAdmin = isAdmin;
	}

	public String getSalutation() {
		return salutation;
	}

	public String getFirstName() {
		return firstName;
	}

	public String getFamilyName() {
		return familyName;
	}

	public Date getDateOfBirth() {
		return dateOfBirth;
	}

	public String getStreet() {
		return street;
	}

	public String getCity() {
		return city;
	}

	public String getZip() {
		return zip;
	}

	public String getPrivateEmail() {
		return privateEmail;
	}

	/**
	 * Return list of group memberships. Is never null, but can be empty list.
	 * @return
	 */
	public List<GroupMembership> getGroupMemberships() {
		return groupMemberships;
	}
	
	@JsonProperty("integrationDosbGender")
	private void setDosbGender(String key) {
		dosbGender = MemberDosbGender.findByKey(key);
	}

	public MemberDosbGender getDosbGender() {
		return dosbGender;
	}

	@JsonProperty("memberGroups")
	private void setGroupMemberships(List<GroupMembership> groupMemberships) {
		this.groupMemberships = groupMemberships;
	}

	@JsonProperty("contactDetails")
	private void flattenContactDetails(Map<String, Object> contactDetails) {
		firstName = (String) contactDetails.get("firstName");
		familyName = (String) contactDetails.get("familyName");
		street = (String) contactDetails.get("street");
		city = (String) contactDetails.get("city");
		zip = (String) contactDetails.get("zip");
		privateEmail = (String) contactDetails.get("privateEmail");
		salutation = (String) contactDetails.get("salutation");

		String dob = (String) contactDetails.get("dateOfBirth");
		try {
			dateOfBirth = dob != null ? dateOnlyFormat.parse(dob) : null;
		} catch (ParseException e) {
			dateOfBirth = null;
			logger.warn("Cannot parse date of Member. Value={}", dob);
		}
	}
	
	@Override
	public String toString() {
		return String.format("Member[id=%s, membershipNumber=%s firstName=%s, familyName=%s]", id, membershipNumber,
				firstName, familyName);
	}

}

File: src/TgoeSrv/Member/Member.php
<?php
declare(strict_types = 1);
namespace TgoeSrv\Member;

use TgoeSrv\Member\Enums\MemberDosbGender;
use TgoeSrv\Member\Enums\DosbSport;

class Member implements \Stringable
{

    public const easyvereinQueryString = '{id,membershipNumber,joinDate,resignationDate,emailOrUserName,_isChairman,contactDetails{salutation,firstName,familyName,dateOfBirth,street,city,zip,privateEmail},memberGroups{memberGroup' . MemberGroup::easyvereinQueryString . '},integrationDosbSport,integrationDosbGender}';

    public const easyvereinDefaultOrder = "membershipNumber";

    private int $id = - 1;

    private string $membershipNumber = '';

    private ?int $joinDate = null;

    private ?int $resignationDate = null;

    private String $loginName = '';

    private bool $isAdmin = false;

    private string $salutation = '';

    private string $firstName = '';

    private string $familyName = '';

    private ?int $dateOfBirth = null;

    private string $street = '';

    private string $city = '';

    private string $zip = '';

    private string $privateEmail = '';

    private MemberDosbGender $dosbGender = MemberDosbGender::UNKNOWN;

    private array $dosbSport = array();

    private array $memberGroups = array();

    public function __construct(?array $arr = null)
    {
        if ($arr != null) {
            $this->id = $arr['id'];
            $this->membershipNumber = $arr['membershipNumber'];
            $this->loginName = $arr['emailOrUserName'];
            $this->isAdmin = $arr['_isChairman'];

            $this->firstName = $arr['contactDetails']['firstName'];
            $this->familyName = $arr['contactDetails']['familyName'];
            $this->street = $arr['contactDetails']['street'];
            $this->city = $arr['contactDetails']['city'];
            $this->zip = $arr['contactDetails']['zip'];
            $this->privateEmail = $arr['contactDetails']['privateEmail'];
            $this->salutation = $arr['contactDetails']['salutation'];

            if (is_array($arr['memberGroups']))
            {
                foreach ($arr['memberGroups'] as $membership) {
                    $g = new MemberGroup($membership['memberGroup']);
                    $this->memberGroups[] = $g;
                }
            }

            if (isset($arr['joinDate']))
                $this->joinDate = strtotime($arr['joinDate']); // ISO 8601 date

            if (isset($arr['resignationDate']))
                $this->resignationDate = strtotime($arr['resignationDate']); // ISO 8601 date

            if (isset($arr['contactDetails']['dateOfBirth']))
                $this->dateOfBirth = strtotime($arr['contactDetails']['dateOfBirth']); // date format yyyy-MM-dd

            $this->dosbGender = MemberDosbGender::findByKey($arr['integrationDosbGender']);

            foreach ($arr['integrationDosbSport'] as $sportString) {
                $sportString = strval($sportString);
                if (strlen($sportString) >= 3) {
                    $sport = DosbSport::findByKey(substr($sportString, 0, 3));
                    if (isset($sport))
                        $this->dosbSport[$sport->getKey()] = $sport;
                }
            }
        }
    }

    public function __toString(): string
    {
        return "Member[id=\"{$this->id}\", membershipNumber=\"{$this->membershipNumber}\", firstName=\"{$this->firstName}\", familyName=\"{$this->familyName}\"]";
    }

    /**
     *
     * @return number
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     *
     * @return string
     */
    public function getMembershipNumber()
    {
        return $this->membershipNumber;
    }

    /**
     *
     * @return number
     */
    public function getJoinDate()
    {
        return $this->joinDate;
    }

    /**
     *
     * @return number
     */
    public function getResignationDate()
    {
        return $this->resignationDate;
    }

    /**
     *
     * @return string
     */
    public function getLoginName()
    {
        return $this->loginName;
    }

    /**
     *
     * @return boolean
     */
    public function isIsAdmin()
    {
        return $this->isAdmin;
    }

    /**
     *
     * @return string
     */
    public function getSalutation()
    {
        return $this->salutation;
    }

    /**
     *
     * @return string
     */
    public function getFirstName()
    {
        return $this->firstName;
    }

    /**
     *
     * @return string
     */
    public function getFamilyName()
    {
        return $this->familyName;
    }

    /**
     *
     * @return number
     */
    public function getDateOfBirth()
    {
        return $this->dateOfBirth;
    }

    /**
     *
     * @return string
     */
    public function getStreet()
    {
        return $this->street;
    }

    /**
     *
     * @return string
     */
    public function getCity()
    {
        return $this->city;
    }

    /**
     *
     * @return string
     */
    public function getZip()
    {
        return $this->zip;
    }

    /**
     *
     * @return string
     */
    public function getPrivateEmail()
    {
        return $this->privateEmail;
    }

    /**
     *
     * @return \TgoeSrv\Member\Enums\MemberDosbGender
     */
    public function getDosbGender()
    {
        return $this->dosbGender;
    }

    /**
     *
     * @return array
     */
    public function getDosbSport()
    {
        return $this->dosbSport;
    }

    /**
     *
     * @return array
     */
    public function getMemberGroups()
    {
        return $this->memberGroups;
    }

    /**
     * Combine first name and family name delimited by space.
     * Returns empty string
     * in case both name parts are empty. Never returns null.
     *
     * @return string
     */
    public function getFullName(): string
    {
        return $this->firstName + (! empty($this->firstName) ? ' ' : '') + $this->familyName;
    }
}

